Model 4 artifact

tylerdy opened the pull request:
## Model 4 artifact
<!-- Ideally, <=50 chars. 50 chars is here..: -->

### Description
<!-- For use in commit message, wrap at 72 chars. 72 chars is here: -->
- *Category*: data artifact
- *JIRA issue*: https://jira.ihme.washington.edu/browse/SSCI-2387
- *Research reference*: https://vivarium-research.readthedocs.io/en/latest/models/causes/alzheimers/presymptomatic_and_mci_gbd_2021/index.html

### Changes and notes
<!-- 
Change description – why, what, anything unexplained by the above.
Include guidance to reviewers if changes are complex.
--> 
- Adds new Alzheimers component keys and data for model 4
- Removes Alz prevalence scale factor, adds scaling factor key to Population component calculated as Alz all states prevalence per attention box equation 1

Open items:
- BBBM incidence math from Nathaniel
- MCI_TO_AD_HAZARD value (constant)
-  Abie suggested naming all keys that are for AD dementia state with "ALZHEIMERS.AD_DEMENTIA", such as ALZHEIMERS.AD_DEMENTIA_INCIDENCE_RATE" which is currently "ALZHEIMERS.PREVALENCE", but this is most of the keys and would make them pretty long. Note that for BBBM and MCI states, the keys are labelled this way to specify the state, eg "ALZHEIMERS.BBBM_INCIDENCE_RATE", while for AD dementia state the state is currently just implied by "ALZHEIMERS.".


### Verification and Testing
<!--
Details on how code was verified. Consider: plots, images, (small) csv files.
-->

albrja commented on src/vivarium_csu_alzheimers/constants/data_values.py:
```diff
@@ -32,3 +33,21 @@
+
+
+# Gamma distribution parameters from Abie's nb
+mean = 3.75  # Fix mean at midpoint of interval [3.5, 4]
```
I think you can skip some of these intermediate variables like mean and variance. Having the infor in a comment like you already do should be fine to calculate GAMMA_SHAPE.

albrja commented on src/vivarium_csu_alzheimers/constants/data_values.py:
```diff
@@ -32,3 +33,21 @@
+
+# Use method of moments to get shape and rate parameters
+GAMMA_SHAPE = mean**2 / variance  # shape parameter alpha
+rate = mean / variance  # rate parameter lambda
```
Make all caps: RATE.

albrja commented on src/vivarium_csu_alzheimers/constants/data_values.py:
```diff
@@ -32,3 +33,21 @@
+# Convert rate to scale for scipy
+GAMMA_SCALE = 1 / rate
+
+GAMMA_DIST = scipy.stats.gamma(GAMMA_SHAPE, scale=GAMMA_SCALE)
```
Nit: Add to this name so it is more descriptive. BBBM_GAMMA_DIST or something like that.

albrja reviewed (commented):
Looks good so far. We can talk through this but I think the easiest way to "clean" some of our data that will have NaNs in it will be to drop the under 5 age groups so all of our datasets have the same shape. This way we are consistent in the artifact and we won't have to worry about any processing in the simulation. 

The easiest way to do that is actually to write code in the get_data function itself but I can explain more tomorrow.

tylerdy commented:
@albrja in addition to the under 5 age groups, all keys except the MCI disability weights contain NAs for under 40 due to calling 'load_alzheimers_duration' either directly or indirectly, which divides by zero incidence for age < 40. Would you recommend dropping under 40 age groups as well?

aflaxman commented on src/vivarium_csu_alzheimers/constants/data_keys.py:
```diff
@@ -31,19 +32,22 @@ def log_name(self):
-    )
-    # Prevalence will just be set to 1 so all simulants are created with the disease
     PREVALENCE: str = "cause.alzheimers_disease_and_other_dementias.prevalence"
+    BBBM_CONDITIONAL_PREVALANCE: str = "cause.bbbm.prevalence"
```
I find it confusing that we have so many version of these strings --- is there a convention we can adopt that will make the all-caps stuff match the double-quoted stuff more precisely?  And should the convention require that the term after "cause" is "alzheimers" if this is all in the `__Alzheimers` class?

albrja replied:
Same with MCI keys in this class.

aflaxman commented on src/vivarium_csu_alzheimers/constants/data_keys.py:
```diff
@@ -31,19 +32,22 @@ def log_name(self):
+    BBBM_CONDITIONAL_PREVALANCE: str = "cause.bbbm.prevalence"
+    MCI_CONDITIONAL_PREVALENCE: str = "cause.mci.prevalence"
     INCIDENCE_RATE: str = "cause.alzheimers_disease_and_other_dementias.incidence_rate"
+    STATE_INCIDENCE_RATE: str = (
```
I think this name is confusing, and hard to match to the `i_AD` name in the docs

aflaxman commented on src/vivarium_csu_alzheimers/constants/data_values.py:
```diff
@@ -32,3 +34,24 @@
+# Convert rate to scale for scipy
+GAMMA_SCALE = 1 / RATE
+
+GAMMA_GAMMA_DIST = scipy.stats.gamma(GAMMA_SHAPE, scale=GAMMA_SCALE)
```
consider naming this to make it clear that it is  specifically for the BBBM transition rate

aflaxman commented on src/vivarium_csu_alzheimers/constants/data_values.py:
```diff
@@ -32,3 +34,24 @@
+
+DW_BBBM = 0
+EMR_BBBM = 0
+EMC_MCI = 0
```
```suggestion
EMR_MCI = 0
```

aflaxman commented on src/vivarium_csu_alzheimers/data/loader.py:
```diff
@@ -58,16 +59,30 @@ def get_data(
     }
-    return mapping[lookup_key](lookup_key, location, years)
+    mapped_value = mapping[lookup_key](lookup_key, location, years)
+    # drop age_groups under 5
```
```suggestion

    # To avoid issues with irrelevant differences between the very young age
    # groups in GBD and FHS data, we will drop rows with age_start < 5 from
    # all age-specific DataFrames
```

aflaxman commented on src/vivarium_csu_alzheimers/data/loader.py:
```diff
@@ -240,3 +245,128 @@ def reshape_to_vivarium_format(df, location):
+    all_state_dur = load_alzheimers_duration_all_states(None, location, years)
+    alz_dur = load_alzheimers_duration(None, location, years)
+    alz_prev = get_data(data_keys.ALZHEIMERS.PREVALENCE, location, years)
+    return ((all_state_dur * alz_prev) / alz_dur).bfill(limit=2).fillna(0)
```
add a comment explaining what `bfill(limit=2)` is intended to accomplish, in case we forget

aflaxman commented on src/vivarium_csu_alzheimers/data/loader.py:
```diff
@@ -240,3 +245,128 @@ def reshape_to_vivarium_format(df, location):
+    return ((all_state_dur * alz_prev) / alz_dur).bfill(limit=2).fillna(0)
+
+
+def load_alzheimers_duration(
```
consider linking to the definition in the docs, and consider using more descriptive names --- this is Delta_AD or something in the docs, right?

aflaxman commented on src/vivarium_csu_alzheimers/data/loader.py:
```diff
@@ -240,3 +245,128 @@ def reshape_to_vivarium_format(df, location):
+    new_bbbm_people = pd.DataFrame(0, index=pop.index, columns=pop.columns)
+    for index, row in new_bbbm_people.iterrows():
+        # create tuples to index population for future age groups and year
+        next_group_future_year = tuple(map(add, list(index), ["", "", W, W, DUR, DUR]))
```
What is going on with the "", "" on this list?

tylerdy replied:
This is the way I thought of to transform the index tuple for one age/sex/year to another - add the values in the original tuple to a "transformation" tuple. Adding empty strings to the index tuple values for location and sex keep those values the same, eg "United States" -> "United States".

I'll make a comment explaining this better, but open to ideas for cleaner ways to do select the necessary indices from the index we are calculating new_bbbm_people for.

aflaxman commented on src/vivarium_csu_alzheimers/data/loader.py:
```diff
@@ -240,3 +245,128 @@ def reshape_to_vivarium_format(df, location):
+            min(2051, scnd_group_future_year[5]),
+        )
+        # get age group incidences
+        # TBD should new_bbbm_people use inc draws? currently single value
```
I think it must use draws!

aflaxman commented on src/vivarium_csu_alzheimers/data/loader.py:
```diff
@@ -240,3 +245,128 @@ def reshape_to_vivarium_format(df, location):
+def load_mci_disability_weight(
+    key: str, location: str, years: int | str | list[int] | None = None
+) -> pd.DataFrame:
+    data = pd.read_csv("/ihme/epi/disability_weights/standard/dw_full.csv")
```
perhaps we should make a copy of this csv, so that we know we can always find it

aflaxman commented on src/vivarium_csu_alzheimers/data/loader.py:
```diff
@@ -240,3 +245,128 @@ def reshape_to_vivarium_format(df, location):
+    # for now, assume csmr is the same for all years based on docs
+    csmr_all_years = pd.DataFrame(csmr, index=acmr.index)
+    mort_MCI = acmr - csmr_all_years  # note that emr_MCI is 0
+    return (1 / data_values.MCI_AVG_DURATION) - mort_MCI
```
FIXME: this results in some negative values for oldest age groups, so we expect the sim will fail to validate due to missing incidence count targets for these old age simulants

tylerdy commented:
Updated code and re-made artifacts to address these comments, including dropping location and incorporating draws from model 3.1 for susceptible_to_bbbm_transition_count, fixing mci_dw column names, and renaming keys and functions and adding documentation. 

Renamed keys only use the full "alzheimers_disease_and_other_dementias" cause if they use load_standard_data. Other keys use "alzheimers" cause. State transition rates/counts are renamed accordingly instead of being named as incidence rates/counts.

New artifacts are in /mnt/team/simulation_science/pub/models/vivarium_csu_alzheimers/artifacts/model4.0

albrja reviewed (approved):
I think this looks good. I have also made a few updates on my branch for this and model 4.1 artifacts in another branch so we should probably just get this in at this point and any other updates we need can go in a separate pr.
